hack/oci-pin/internal/sums/sums.go

```diff
@@ -0,0 +1,225 @@
+	normalizedRefs, err := normalizeRefs(refs)
+	if err != nil {
+		return err
+	}
+	for _, ref := range normalizedRefs {
+		if _, err := fmt.Fprintf(out, "%s  %s\n", ref.Sum(), ref.Reference()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Merge merges multiple sha256sums files into one.
+func Merge(refs [][]PinnedImageReference, out io.Writer) error {
+	mergedRefs := make([]PinnedImageReference, 0)
+	for _, refs := range refs {
+		mergedRefs = append(mergedRefs, refs...)
+	}
+	return Create(mergedRefs, out)
+}
+
+// Parse parses a sha256sums file.
+func Parse(in io.Reader) ([]PinnedImageReference, error) {
+	scanner := newLineScanner(in)
+	return parseLines(scanner)
+}
+
+// PinnedImageReference contains the components of a pinned image reference.
+type PinnedImageReference struct {
+	Registry string
+	Prefix   string
+	Name     string
+	Tag      string
+	Digest   string
+}
+
+// Reference returns the string representation of the reference (without the digest).
+func (r PinnedImageReference) Reference() string {
+	var base string
+	if r.Prefix == "" {
+		base = path.Join(r.Registry, r.Name)
+	} else {
+		base = path.Join(r.Registry, r.Prefix, r.Name)
+	}
+	var tag string
+	if r.Tag != "" {
+		tag = ":" + r.Tag
+	}
+	return base + tag
+}
+
+// Sum returns the string representation of the digest.
+// The digest is expected to be in the oci format "sha256:hex".
+// The resulting Sum is only the hex part.
+func (r PinnedImageReference) Sum() string {
+	return r.Digest[len("sha256:"):]
+}
+
+// normalizeRefs normalizes the image references.
+// It sorts the references and removes duplicates.
+// If conflicting digests are found, an error is returned.
+func normalizeRefs(refs []PinnedImageReference) ([]PinnedImageReference, error) {
+	sortRefs(refs)
+	uniqueRefs := make([]PinnedImageReference, 0, len(refs))
+	var prev PinnedImageReference
+	for _, ref := range refs {
+		equal, err := compareRefs(prev, ref)
+		if err != nil {
+			return nil, err
+		}
+		if !equal {
+			uniqueRefs = append(uniqueRefs, ref)
+			prev = ref
+		}
+	}
+	return uniqueRefs, nil
+}
+
+func sortRefs(refs []PinnedImageReference) {
+	sort.Slice(refs, func(i, j int) bool {
+		if refs[i].Registry != refs[j].Registry {
+			return refs[i].Registry < refs[j].Registry
+		}
+		if refs[i].Prefix != refs[j].Prefix {
+			return refs[i].Prefix < refs[j].Prefix
+		}
+		if refs[i].Name != refs[j].Name {
+			return refs[i].Name < refs[j].Name
+		}
+		if refs[i].Tag != refs[j].Tag {
+			return refs[i].Tag < refs[j].Tag
+		}
+		return refs[i].Digest < refs[j].Digest
+	})
+}
+
+func compareRefs(a, b PinnedImageReference) (equal bool, err error) {
+	if a.Registry != b.Registry {
+		return false, nil
+	}
+	if a.Prefix != b.Prefix {
+		return false, nil
+	}
+	if a.Name != b.Name {
+		return false, nil
+	}
+	if a.Tag != b.Tag {
+		return false, nil
+	}
+	if a.Digest != b.Digest {
+		return false, errors.New("conflicting digests")
```

Comment:
I'm not 100% convinced about throwing an error in this case. Viewing the function alone, it does not make sense to error in this case, since we can successfully determine, that the refs do not match. 

But given the usage of this function it makes more sense and I don't know what to change. Maybe add a comment to this function?